### PR TITLE
Add Cardano currency support

### PR DESCRIPTION
**Pull request by ddeath**
This PR is for adding Cardano currency support and is intended to merge with https://github.com/trezor/trezor-core/pull/250

For the support of Cardano currency was required to add cardano-specific ed25519 crypto support.
The crypto code is based on [cardano-crypto](https://github.com/input-output-hk/cardano-crypto) and [cardano-sl](https://github.com/input-output-hk/cardano-sl) - official Cardano repos.

We are NOT official developers of Cardano, however we are developers of [www.cardanolite.com](https://www.cardanolite.com)

### CODE REVIEW

**Comment by prusnak** [comment]
Do not merge unless https://github.com/trezor/trezor-common/pull/148 is merged first

**Comment by prusnak** [comment]
@ddeath you have lots of useless whitespace changes in the code, please remove these whitespace changes

**Comment by ddeath** [comment]
@prusnak fixed

**Comment by prusnak** [comment]
Can you please squash the commit into one? Now we have two big commits changing whitespace which is making review difficult as we want to focus just on the functional changes.

**Comment by prusnak** [comment]
It seems that you never set `private_key_extension` to a particular value except setting it to zero via `memzero`. That does not look right. Is that correct?

**Comment by ddeath** [comment]
Yes it is correct. The private key extension is set via private key. This is because some functions works with private key defined as 32-byte and some as 64-byte. So for compatibility reasons we solved that by adding private key extension right after private key in HDnode stuct.

This is because how cardano works. You can read about it [here](https://cardanolaunch.com/assets/Ed25519_BIP.pdf) on page 2, section IV and point 1)

Also going to add the tests.

File: ed25519-donna/ed25519.h [hunk 1]
```diff
@@ -14,9 +14,15 @@ typedef unsigned char curve25519_key[32];
 typedef unsigned char ed25519_cosi_signature[32];
 
 void ed25519_publickey(const ed25519_secret_key sk, ed25519_public_key pk);
+#ifdef USE_CARDANO
+void ed25519_publickey_cardano(const ed25519_secret_key sk, const ed25519_secret_key skext, ed25519_public_key pk);
```
Comment (prusnak) [inline ed25519-donna/ed25519.h]: Let's rename this to `ed25519_publickey_ext`

File: ed25519-donna/ed25519.h [hunk 2]
```diff
@@ -14,9 +14,15 @@ typedef unsigned char curve25519_key[32];
...
+#endif
 
 int ed25519_sign_open(const unsigned char *m, size_t mlen, const ed25519_public_key pk, const ed25519_signature RS);
 void ed25519_sign(const unsigned char *m, size_t mlen, const ed25519_secret_key sk, const ed25519_public_key pk, ed25519_signature RS);
+#ifdef USE_CARDANO
+void ed25519_sign_cardano(const unsigned char *m, size_t mlen, const ed25519_secret_key sk, const ed25519_secret_key skext, const ed25519_public_key pk, ed25519_signature RS);
```
Comment (prusnak) [inline ed25519-donna/ed25519.h]: Let's rename this to `ed25519_sign_ext`

File: bip32.h [hunk 3]
```diff
@@ -45,7 +45,12 @@ typedef struct {
 	uint32_t depth;
 	uint32_t child_num;
 	uint8_t chain_code[32];
+
+	// the private key and private key extension must be in this order and next to each other
+	// because some of the ed25519 functions works with 32-byte and some with 64-byte private key
 	uint8_t private_key[32];
+	uint8_t private_key_extension[32];
```
Comment (prusnak) [inline bip32.h]: TBH I don't really like this. Most of the code check analysis tools will scream anyway when you'll try to write 64 bytes into private_key field.

I suggest to move `private_key_extension` AFTER `public_key` field. Then fix all broken stuff (it should be quite obvious now). Then return it back, but without the comment.

From what I can tell fixing the issues should be quite easy, just split 

`memcpy(priv_key,inout->private_key,32);` into

```
memcpy(priv_key,inout->private_key,64);
memcpy(priv_key+32,inout->private_key_extension,32);
```

and

`memcpy(inout->private_key, res_key, 64);` into

```
memcpy(inout->private_key, res_key, 32);
memcpy(inout->private_key_extension, res_key + 32, 32);
```

+ maybe couple of more similar things.

File: bip32.c [hunk 4]
```diff
@@ -252,6 +268,194 @@ int hdnode_private_ckd(HDNode *inout, uint32_t i)
...
+	uint8_t entropy[66];
+	memzero(entropy,66);
+	bitcount = mnemonic_to_entropy(mnemonic,entropy+2);
+	if (bitcount == 0){
+		return -1;
+	}
+	int entlength = bitcount / 8;
+	uint8_t hash[32];
+	uint8_t cbor[32+2];
+	if(entlength < 24){
```
Comment (prusnak) [inline bip32.c]: `memmove` is ugly, lets rewrite this if block + black2b call below into:

```
	if (entlength < 24){
		entropy[1] = 64 | entlength; // cbor encodes length directly into first byte if its smaller than 24
		blake2b(entropy + 1,entlength + 1,hash, 32);
	} else {
		entropy[0] = 88;
		entropy[1] = entlength;
		blake2b(entropy,entlength + 2,hash, 32);
	}
```

File: bip32.c [hunk 5]
```diff
@@ -252,6 +268,194 @@ int hdnode_private_ckd(HDNode *inout, uint32_t i)
...
+    int i = 30;
+    
+    for(; val && i ; --i, val /= 10){
+        buf[i] = "0123456789"[val % 10];
+    }
+    
+    memcpy(out,&buf[i+1],strlen(&buf[i+1])+1);    
+}
+
+int hdnode_from_mnemonic_cardano(HDNode *out, const char *mnemonic){
```
Comment (prusnak) [inline bip32.c]: Let's change this function (so it matches the API of `hdnode_from_seed`) into:

``` c
int hdnode_from_seed_cardano(const uint8_t *seed, int seed_len, HDNode *out);
```

One will then use the following to get the same functionality:

```
int entropy_len = mnemonic_to_entropy(mnemonic, entropy);
hdnode_from_seed_cardano(entropy, entropy_len / 8, &out);
```

**Comment by ddeath** [comment]
The suggested changes are implemented now. We will commit test later today, however the code is working after changes as I confirmed it by running tests in trezor-core repository.

**Comment by prusnak** [comment]
Great! Looking forward to the tests! Are there any official test vectors? I don't see any in the BIP32-Ed25519 paper. Please use the official ones if there are any in your tests.

**Comment by ddeath** [comment]
We added tests for derivation schema 1 from [here](https://github.com/input-output-hk/cardano-crypto/blob/master/tests/goldens/cardano/crypto/wallet/BIP39-128).

**Comment by ddeath** [comment]
Just wanted ask if the test vectors from [cardano repository](https://github.com/input-output-hk/cardano-crypto/blob/master/tests/goldens/cardano/crypto/wallet/BIP39-128) are enough

**Comment by prusnak** [comment]
Yes, it's fine. I'll review this soon.

**Comment by prusnak** [comment]
Thank you!

**Pull request merged**